### PR TITLE
ES6: migrated "entity/message" 

### PR DESCRIPTION
**Pull request by bennyn**

### CODE REVIEW

File: app/script/entity/message/Asset.js [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+z.entity.Asset = class Asset {
+  constructor(id) {
+    this.id = id;
+    this.key = '';
+    this.type = '';
+  }
+
+  /**
+   * Check if asset is a medium image.
+   *
```
Comment (gregor) [inline app/script/entity/message/Asset.js]: Do we need all these empty lines in docs?

File: app/script/entity/message/Asset.js [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+    return this.type === z.assets.AssetType.LOCATION;
+  }
+
+  /**
+   * Check if asset is a video.
+   *
+   * @returns {Boolean} Is asset of type video
+   */
+  is_video() {
+    const is_video_asset = (this.type === z.assets.AssetType.FILE) && (this.file_type != null ? this.file_type.startsWith('video') : undefined);
```
Comment (gregor) [inline app/script/entity/message/Asset.js]: `const is_video_asset = this.type === z.assets.AssetType.FILE && this.file_type && this.file_type.startsWith('video');` 

seems easier to read to me

File: app/script/entity/message/Asset.js [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+    return false;
+  }
+
+  /**
+   * Check if asset is a audio.
+   *
+   * @returns {Boolean} Is asset of type audio
+   */
+  is_audio() {
+    if ((this.type === z.assets.AssetType.FILE) && (this.file_type != null ? this.file_type.startsWith('audio') : undefined)) {
```
Comment (gregor) [inline app/script/entity/message/Asset.js]: see above

File: app/script/entity/message/ContentMessage.js [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+      .map((user_et) => user_et.first_name()))
+      .join(', ')
+    );
+
+    this.display_edited_timestamp = () => {
+      return z.localization.Localizer.get_text({
+        id: z.string.conversation_edit_timestamp,
+        replace: {
+          placeholder: '%@timestamp',
+          content: moment(this.edited_timestamp).format('HH:mm'),
```
Comment (gregor) [inline app/script/entity/message/ContentMessage.js]: eslint will flag this

File: app/script/entity/message/ContentMessage.js [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+
+    this.assets = ko.observableArray([]);
+    this.nonce = null;
+    this.super_type = z.message.SuperType.CONTENT;
+    this.replacing_message_id = null;
+    this.edited_timestamp = null;
+
+    this.reactions = ko.observable({});
+    this.reactions_user_ets = ko.observableArray();
+    this.reactions_user_ids = ko.pureComputed(() => (this.reactions_user_ets()
```
Comment (gregor) [inline app/script/entity/message/ContentMessage.js]:
````
this.reactions_user_ids = ko.pureComputed(() => {
  this.reactions_user_ets()
    .map((user_et) => user_et.first_name())
    .join(', ')
});
````

File: app/script/entity/message/ContentMessage.js [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+      },
+    });
+    this.other_likes = ko.pureComputed(() => {
+      return this.reactions_user_ets().filter((user_et) => !user_et.is_me);
+    });
+    this.show_likes = ko.observable(false);
+
+    this.like_caption = ko.pureComputed(() => {
+      if (this.reactions_user_ets().length <= 5) {
+        return (this.reactions_user_ets()
```
Comment (gregor) [inline app/script/entity/message/ContentMessage.js]:
````
return (this.reactions_user_ets()
  .map((user_et) => user_et.first_name())
  .join(', ');
````

File: app/script/entity/message/ContentMessage.js [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+  }
+
+  /**
+   * Download message content.
+   * @returns {undefined} No return value
+   */
+  download() {
+    const asset_et = this.get_first_asset();
+    const file_name = this.get_content_name();
+    return asset_et.download(file_name);
```
Comment (gregor) [inline app/script/entity/message/ContentMessage.js]: Not expected to return a value: `asset_et.download(file_name);`

File: app/script/entity/message/File.js [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+      return undefined;
+    });
+
+    this.upload_id = ko.observable();
+    this.upload_progress = ko.observable();
+    this.uploaded_on_this_client = ko.observable(false);
+    this.upload_failed_reason = ko.observable();
+    this.upload_cancel = undefined;
+    this.pending_upload = ko.pureComputed(() => {
+      return (this.status() === z.assets.AssetTransferState.UPLOADING) && this.uploaded_on_this_client();
```
Comment (gregor) [inline app/script/entity/message/File.js]: `return this.status() === z.assets.AssetTransferState.UPLOADING && this.uploaded_on_this_client();`

File: app/script/entity/message/File.js [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+   * Loads and decrypts otr asset preview
+   *
+   * @returns {Promise} Returns a promise that resolves with the asset as blob
+   */
+  load_preview() {
+    if (this.preview_resource()) {
+      return this.preview_resource().load();
+    }
+
+    return undefined;
```
Comment (gregor) [inline app/script/entity/message/File.js]: It if is supposed to return a promise we should not return undefined but Promise.resolve or Promise.reject. Maybe a return value is not needed though.

File: app/script/entity/message/File.js [hunk 10]
```diff
@@ -0,0 +1,176 @@
...
+          this.status(z.assets.AssetTransferState.UPLOADED);
+          return blob;
+        })
+        .catch((error) => {
+          this.status(z.assets.AssetTransferState.UPLOADED);
+          throw error;
+        });
+    }
+
+    return undefined;
```
Comment (gregor) [inline app/script/entity/message/File.js]: See above

File: app/script/entity/message/File.js [hunk 11]
```diff
@@ -0,0 +1,176 @@
...
+  }
+
+  /**
+   * Loads and decrypts otr asset as initiates download
+   *
+   * @returns {Promise} Returns a promise that resolves with the asset as blob
+   */
+  download() {
+    if (this.status() !== z.assets.AssetTransferState.UPLOADED) {
+      return;
```
Comment (gregor) [inline app/script/entity/message/File.js]: And once more

File: app/script/entity/message/File.js [hunk 12]
```diff
@@ -0,0 +1,176 @@
...
+  }
+
+  cancel_download() {
+    this.status(z.assets.AssetTransferState.UPLOADED);
+
+    if (this.original_resource()) {
+      return this.original_resource().cancel_download();
+    }
+
+    return undefined;
```
Comment (gregor) [inline app/script/entity/message/File.js]: Promise or no promise is the question

File: app/script/entity/message/File.js [hunk 13]
```diff
@@ -0,0 +1,176 @@
...
+  }
+
+  cancel(message_et) {
+    if (typeof this.upload_cancel === 'function') {
+      this.upload_cancel();
+    }
+    amplify.publish(z.event.WebApp.CONVERSATION.ASSET.CANCEL, message_et);
+    return amplify.publish(z.event.WebApp.ANALYTICS.EVENT, z.tracking.EventName.FILE.UPLOAD_CANCELLED, {
```
Comment (gregor) [inline app/script/entity/message/File.js]: Definitely not something we should to return

File: app/script/entity/message/File.js [hunk 14]
```diff
@@ -0,0 +1,176 @@
...
+      size_bytes: this.file_size,
+      size_mb: z.util.bucket_values((this.file_size / 1024 / 1024), [0, 5, 10, 15, 20, 25]),
+      type: z.util.get_file_extension(this.file_name),
+    }
+    );
+  }
+
+  reload() {
+    return this.logger.info('Restart upload');
```
Comment (gregor) [inline app/script/entity/message/File.js]: What's this? Do not return but not implemented or not needed for this type of asset.

File: app/script/entity/message/MediumImage.js [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+   */
+  download(file_name) {
+    if (this.resource()) {
+      return this.resource().load()
+        .then(function(blob) {
+          z.util.download_blob(blob, file_name);
+        });
+    }
+
+    return undefined;
```
Comment (gregor) [inline app/script/entity/message/MediumImage.js]: Promise or not?

File: app/script/entity/message/MemberMessage.js [hunk 16]
```diff
@@ -0,0 +1,156 @@
...
+
+    this.super_type = z.message.SuperType.MEMBER;
+    this.member_message_type = z.message.SystemMessageType.NORMAL;
+
+    this.user_ets = ko.observableArray();
+    this.user_ids = ko.observableArray();
+
+    // Users joined the conversation without sender
+    this.joined_user_ets = ko.pureComputed(() => {
+      return (Array.from(this.user_ets())
```
Comment (gregor) [inline app/script/entity/message/MemberMessage.js]: Observable arrays are always arrays - no need for array.from
````
  return this.user_ets()
    .filter((user_et) => user_et.id !== this.user().id)
    .map((user_et) => user_et);
````

File: app/script/entity/message/MemberMessage.js [hunk 17]
```diff
@@ -0,0 +1,156 @@
...
+        .filter((user_et) => user_et.id !== this.user().id)
+        .map((user_et) => user_et)
+      );
+    });
+
+    // Users joined the conversation without self
+    this.remote_user_ets = ko.pureComputed(() => {
+      return (Array.from(this.user_ets())
```
Comment (gregor) [inline app/script/entity/message/MemberMessage.js]: Same as above:

````
  return this.user_ets()
    .filter((user_et) => !user_et.is_me)
    .map((user_et) => user_et);
````

File: app/script/entity/message/MemberMessage.js [hunk 18]
```diff
@@ -0,0 +1,156 @@
...
+        },
+      });
+    };
+
+    this.show_large_avatar = () => {
+      const large_avatar_types = [
+        z.message.SystemMessageType.CONNECTION_ACCEPTED,
+        z.message.SystemMessageType.CONNECTION_REQUEST,
+      ];
+      return Array.from(large_avatar_types).includes(this.member_message_type);
```
Comment (gregor) [inline app/script/entity/message/MemberMessage.js]: No need for array.from on array

`return large_avatar_types.includes(this.member_message_type);`

File: app/script/entity/message/MemberMessage.js [hunk 19]
```diff
@@ -0,0 +1,156 @@
...
+            return this._get_caption_with_names(z.string.conversation_member_leave_removed_you);
+          }
+          return this._get_caption_with_names(z.string.conversation_member_leave_removed);
+        case z.event.Backend.CONVERSATION.MEMBER_JOIN:
+          if (this.user().is_me) {
+            return this._get_caption_with_names(z.string.conversation_member_join_you);
+          }
+          return this._get_caption_with_names(z.string.conversation_member_join);
+      }
+    }
```
Comment (gregor) [inline app/script/entity/message/MemberMessage.js]: too much indentation
````
  },
  this, {deferEvaluation: true});
````

File: app/script/entity/message/Message.js [hunk 20]
```diff
@@ -0,0 +1,351 @@
...
+ *
+ */
+
+'use strict';
+
+window.z = window.z || {};
+window.z.entity = z.entity || {};
+
+z.entity.Message = class Message {
+  /**
```
Comment (gregor) [inline app/script/entity/message/Message.js:26]: Proper JSDoc missing

File: app/script/entity/message/Message.js [hunk 21]
```diff
@@ -0,0 +1,351 @@
...
+  constructor(id, super_type) {
+    this.equals = this.equals.bind(this);
+    this.is_expired = this.is_expired.bind(this);
+    this.start_ephemeral_timer = this.start_ephemeral_timer.bind(this);
+    if (id == null) {
+      id = '0';
+    }
+    this.id = id;
+    if (super_type == null) {
+      super_type = '';
```
Comment (gregor) [inline app/script/entity/message/Message.js]: Defaults can be set in params:

````
constructor(id = '0', super_type = '') {
  this.equals = this.equals.bind(this);
  this.is_expired = this.is_expired.bind(this);
  this.start_ephemeral_timer = this.start_ephemeral_timer.bind(this);
  this.id = id;
  this.super_type = super_type;
````

File: app/script/entity/message/Message.js [hunk 22]
```diff
@@ -0,0 +1,351 @@
...
+        return z.message.EphemeralStatusType.INACTIVE;
+      }
+
+      if (_.isString(this.ephemeral_expires())) {
+        if (this.ephemeral_expires() > Date.now()) {
+          return z.message.EphemeralStatusType.ACTIVE;
+        }
+        return z.message.EphemeralStatusType.TIMED_OUT;
+
+      }
```
Comment (gregor) [inline app/script/entity/message/Message.js]: Weird empty line positioning:

````
if (_.isString(this.ephemeral_expires())) {
  if (this.ephemeral_expires() > Date.now()) {
    return z.message.EphemeralStatusType.ACTIVE;
  }

  return z.message.EphemeralStatusType.TIMED_OUT;
}
````

File: app/script/entity/message/Message.js [hunk 23]
```diff
@@ -0,0 +1,351 @@
...
+
+    this.display_timestamp_short = () => {
+      const date = moment.unix(this.timestamp() / 1000);
+      return date.local().format('HH:mm');
+    };
+
+    this.sender_name = ko.pureComputed(() => {
+      return z.util.get_first_name(this.user());
+    }
+      , this, {deferEvaluation: true});
```
Comment (gregor) [inline app/script/entity/message/Message.js]: Indentation:

````
},
this, {deferEvaluation: true});
```

File: app/script/entity/message/Message.js [hunk 24]
```diff
@@ -0,0 +1,351 @@
...
+    return this.id === (message_et != null ? message_et.id : undefined);
+  }
+
+  /**
+   * Check if message contains an asset of type file.
+   * @returns {Boolean} Message contains any file type asset
+   */
+  has_asset() {
+    if (this.is_content()) {
+      for (const asset_et of Array.from(this.assets())) {
```
Comment (gregor) [inline app/script/entity/message/Message.js]: assets is an array:

`for (const asset_et of this.assets()) {`

File: app/script/entity/message/Message.js [hunk 25]
```diff
@@ -0,0 +1,351 @@
...
+    return false;
+  }
+
+  /**
+   * Check if message contains a file asset.
+   * @returns {Boolean} Message contains a file
+   */
+  has_asset_file() {
+    if (this.is_content()) {
+      for (const asset_et of Array.from(this.assets())) {
```
Comment (gregor) [inline app/script/entity/message/Message.js]: `for (const asset_et of this.assets()) {`

File: app/script/entity/message/Message.js [hunk 26]
```diff
@@ -0,0 +1,351 @@
...
+    return false;
+  }
+
+  /**
+   * Check if message contains any image asset.
+   * @returns {Boolean} Message contains any image
+   */
+  has_asset_image() {
+    if (this.is_content()) {
+      for (const asset_et of Array.from(this.assets())) {
```
Comment (gregor) [inline app/script/entity/message/Message.js]: `for (const asset_et of this.assets()) {`

File: app/script/entity/message/Message.js [hunk 27]
```diff
@@ -0,0 +1,351 @@
...
+    return false;
+  }
+
+  /**
+   * Check if message contains a location asset.
+   * @returns {Boolean} Message contains a location
+   */
+  has_asset_location() {
+    if (this.is_content()) {
+      for (const asset_et of Array.from(this.assets())) {
```
Comment (gregor) [inline app/script/entity/message/Message.js]: `for (const asset_et of this.assets()) {`

File: app/script/entity/message/Message.js [hunk 28]
```diff
@@ -0,0 +1,351 @@
...
+    return false;
+  }
+
+  /**
+   * Check if message contains a text asset.
+   * @returns {Boolean} Message contains text
+   */
+  has_asset_text() {
+    if (this.is_content()) {
+      for (const asset_et of Array.from(this.assets())) {
```
Comment (gregor) [inline app/script/entity/message/Message.js]: `for (const asset_et of this.assets()) {`

File: app/script/entity/message/Message.js [hunk 29]
```diff
@@ -0,0 +1,351 @@
...
+    }
+    return false;
+  }
+
+  /**
+   * Check if message contains a nonce.
+   * @returns {Boolean} Message contains a nonce
+   */
+  has_nonce() {
+    return [z.message.SuperType.CONTENT].includes(this.super_type);
```
Comment (gregor) [inline app/script/entity/message/Message.js]: No longer needs an array check:

`return this.super_type === z.message.SuperType.CONTENT` or `return this.is_content()`

File: app/script/entity/message/Message.js [hunk 30]
```diff
@@ -0,0 +1,351 @@
...
+    return ![z.assets.AssetTransferState.DOWNLOADING,
+      z.assets.AssetTransferState.UPLOADING].includes(this.get_first_asset().status());
+  }
+
+  /**
+   * Check if the message content can be downloaded
+   * @returns {Boolean} Is message of type content
+   */
+  is_downloadable() {
+    return ((typeof this.get_first_asset === 'function' ? this.get_first_asset().download : undefined) != null);
```
Comment (gregor) [inline app/script/entity/message/Message.js]: Can we break this up to be more human readable?

File: app/script/entity/message/Message.js [hunk 31]
```diff
@@ -0,0 +1,351 @@
...
+  is_expired() {
+    return this.ephemeral_expires() === true;
+  }
+
+  /**
+   * Check if message can be reacted to.
+   * @returns {Boolean} True, if message type supports reactions.
+   */
+  is_reactable() {
+    return this.is_content() && (this.status() !== z.message.StatusType.SENDING) && !this.is_ephemeral();
```
Comment (gregor) [inline app/script/entity/message/Message.js]: `return this.is_content() && !this.is_ephemeral() && this.status() !== z.message.StatusType.SENDING;`

File: app/script/entity/message/Message.js [hunk 32]
```diff
@@ -0,0 +1,351 @@
...
+        .toString());
+    }
+
+    this.ephemeral_remaining(this.ephemeral_expires() - Date.now());
+
+    this.ephemeral_interval_id = window.setInterval(() => {
+      this.ephemeral_remaining(this.ephemeral_expires() - Date.now());
+      return this.ephemeral_caption(z.util.format_time_remaining(this.ephemeral_remaining()));
+    }
+      , 250);
```
Comment (gregor) [inline app/script/entity/message/Message.js]:
````
},
250};
````

File: app/script/entity/message/Message.js [hunk 33]
```diff
@@ -0,0 +1,351 @@
...
+
+    return this.ephemeral_timeout_id = window.setTimeout(() => {
+      amplify.publish(z.event.WebApp.CONVERSATION.EPHEMERAL_MESSAGE_TIMEOUT, this);
+      return window.clearInterval(this.ephemeral_interval_id);
+    }
```
Comment (gregor) [inline app/script/entity/message/Message.js]:
````
},
this.ephemeral_remaining();
````

File: app/script/entity/message/PingMessage.js [hunk 34]
```diff
@@ -0,0 +1,44 @@
...
+z.entity.PingMessage = class PingMessage extends z.entity.Message {
+  constructor() {
+    super();
+    this.super_type = z.message.SuperType.PING;
+
+    this.caption = ko.pureComputed(() => {
+      const string = this.user().is_me ? z.string.conversation_ping_you : z.string.conversation_ping;
+      return z.localization.Localizer.get_text(string);
+    }
+      , this, {deferEvaluation: true});
```
Comment (gregor) [inline app/script/entity/message/PingMessage.js]:
````
},
this, {deferEvaluation: true});
````

File: app/script/entity/message/Text.js [hunk 35]
```diff
@@ -0,0 +1,60 @@
...
+
+window.z = window.z || {};
+window.z.entity = z.entity || {};
+
+z.entity.Text = class Text extends z.entity.Asset {
+  constructor(id, text) {
+    if (text == null) {
+      text = '';
+    }
+    super(id);
```
Comment (gregor) [inline app/script/entity/message/Text.js]: Default value in params

````
  constructor(id, text = '') {
    super(id);
````

File: app/script/entity/message/File.js [hunk 36]
```diff
@@ -0,0 +1,176 @@
...
+   */
+  download() {
+    if (this.status() !== z.assets.AssetTransferState.UPLOADED) {
+      return;
+    }
+
+    const download_started = Date.now();
+    const tracking_data = {
+      size_bytes: this.file_size,
+      size_mb: z.util.bucket_values((this.file_size / 1024 / 1024), [0, 5, 10, 15, 20, 25]),
```
Comment (lipis) [inline app/script/entity/message/File.js:121]: `MEBIBIT = 1024 * 1024` https://en.wikipedia.org/wiki/Mebibit :D

File: app/script/entity/message/PingMessage.js [hunk 37]
```diff
@@ -0,0 +1,44 @@
...
+
+    this.get_icon_classes = ko.pureComputed(() => {
+      let css_classes = this.accent_color();
+      if ((Date.now() - this.timestamp()) < 2000) {
```
Comment (lipis) [inline app/script/entity/message/PingMessage.js]: const

File: app/script/entity/message/Message.js (hunk 20 above)
Comment (bennycode) [reply app/script/entity/message/Message.js:26]: Fixed return description. :)

File: app/script/entity/message/Message.js [hunk 38]
```diff
@@ -0,0 +1,344 @@
...
+    return ![z.assets.AssetTransferState.DOWNLOADING,
+      z.assets.AssetTransferState.UPLOADING].includes(this.get_first_asset().status());
+  }
+
+  /**
+   * Check if the message content can be downloaded.
+   * @returns {Boolean} True, if the message has downloadable content.
+   */
+  is_downloadable() {
```
Comment (gregor) [inline app/script/entity/message/Message.js]: Old code: `return @get_first_asset?().download?`

Something seems to be missing here

Comment (bennycode) [reply app/script/entity/message/Message.js]: #fixed 😎

File: app/script/entity/message/Message.js [hunk 39]
```diff
@@ -213,7 +213,11 @@ z.entity.Message = class Message {
    */
   is_downloadable() {
     if (typeof this.get_first_asset === 'function') {
-      return true;
+      const asset_et = this.get_first_asset();
+      if (asset_et && (typeof asset_et.download === 'function')) {
+        return true;
+      }
+      return false;
```
Comment (gregor) [inline app/script/entity/message/Message.js]: no need for this extra return, covered by 222

Comment (bennycode) [reply app/script/entity/message/Message.js]: #fixed :)

**Comment by herrmannplatz** [comment]
finally 🎉

**Comment by bennycode** [comment]
Gut Ding will Weile haben. :D

**Pull request merged**